### PR TITLE
Autocomplete unbind

### LINKED ISSUE #285: Autocomplete bindings in :unbind command

### PR DESCRIPTION
**Pull request by keddelzz**
closes #285

### CODE REVIEW

**Comment by rexim** [comment]
@keddelzz that is so cool! Thanks! :) But I'm gonna review that later. Falling asleep already -_-

**Comment by keddelzz** [comment]
>   But I'm gonna review that later.

No problem. :)

File: src/main/scala/me/rexim/morganey/Commands.scala [hunk 1]
```diff
@@ -9,24 +9,31 @@ import scala.util.{Failure, Success}
...
 object Commands {
 
-  final type Command = ReplContext => Computation[ReplResult[String]]
+  type CommandTask = ReplContext => Computation[ReplResult[String]]
 
-  private def command(f: ReplContext => ReplResult[String]): Command =
+  case class Command(
+    name: String,
+    task: String => ReplContext => ReplResult[String],
+    hasTermArgument: Boolean)
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/Commands.scala]: Can we get rid of that? I mean it's no better than having a function with a boolean argument. Why can't it be something like

``` scala
trait Command
case class StringCommand extends Command
case class TermCommand extends Command
```

File: src/main/scala/me/rexim/morganey/ReplAutocompletion.scala [hunk 2]
```diff
@@ -127,18 +130,34 @@ object ReplAutocompletion {
...
-  private object IsCommand {
+  private object SimpleCommand {
 
     def unapply(line: String): Option[List[String]] = {
       val potentialCommand = parseCommand(line).map(_._1)
       val matchingCommands = potentialCommand map { p =>
-        commands.keySet filter (_ startsWith p)
+        commands.values collect {
+          case c if c.name startsWith p =>
+            if (c.hasTermArgument) c.name + " " else c.name
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/ReplAutocompletion.scala]: I don't understand this code. Could you please elaborate on what's going on here? Why do we care about term arguments when we parse a simple command?

**Review by rexim — CHANGES_REQUESTED** [review]

**Review by ForNeVeR — APPROVED** [review]
I have no objections except from what @rexim said.

File: src/main/scala/me/rexim/morganey/ReplAutocompletion.scala (hunk 2 above)
Comment (keddelzz) [reply src/main/scala/me/rexim/morganey/ReplAutocompletion.scala]: My goal with this code, was to autocomplete commands that have a term as an argument to ":command " (notice the space at the end). The space was, as far as I thought, necessary that the autocompletion for commands with arguments (`CommandWithArg`) can distinguish the command name and the arguments.

However, I was wrong. The space isn't needed.

I should have tested that earlier.

**Comment by rexim** [comment]
@keddelzz thanks for the update! I'm gonna review it a bit later today.

File: src/main/scala/me/rexim/morganey/ReplAutocompletion.scala [hunk 3]
```diff
@@ -127,18 +130,39 @@ object ReplAutocompletion {
...
+      cmd match {
+        case StringCommand(_, _) => false
+        case TermCommand(_, _)   => true
+      }
+
+    def unapply(line: String): Option[(String, String)] = {
+      val potentialCommand = parseCommand(line)
+      potentialCommand flatMap { case (p, arg) =>
+        commands.values find { cmd =>
+          cmd.name == p && hasTermArgument(cmd)
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/ReplAutocompletion.scala]: Could you please inline `hasTermArgument` here like

``` scala
commands.values find {
  case TermCommand(name, _) => name == p
  case _ => false
}
```

**Review by rexim — CHANGES_REQUESTED** [review]

**Review by rexim — APPROVED** [review]

**Comment by rexim** [comment]
@keddelzz thank you very much for the changes!
@ForNeVeR thanks for the review!

**Pull request merged**